### PR TITLE
Column pairs

### PR DESCRIPTION
**Pull request by abegong**

### CODE REVIEW

**Comment by abegong** [comment]
In addition to implementing column_pair_expectations, this branch also:

* Defines a convenience method called `run_encapsulated_test` that implements an early version of #167
* Removes the pandas dependency in base.py.
* Changes the format for `summary_obj.partial_exception_counts`. Now it's a list of dictionaries, rather than a dictionary. This allows us to sort entries in descending order.

**Comment by abegong** [comment]
@jpcampb2 , this should be ready to merge---our first column_pair expectations!

Please check the API and see if it makes sense to you. The `keep_missing` syntax doesn't feel very user-friendly to me, but I haven't been able to come up with anything better.

@gouker, Once this merges, this should give you what you need from #194.

**Comment by jpcampb2** [comment]
Beginning to look at this more deeply again:

1. `expect_column_A_values_to_be_greater_than_column_B` doesn't follow the pattern of column_pair_... as a (future) decorator option. Recommend instead `expect_column_pair_values_A_to_be_greater_than_B`

2. `keep_missing` is super confusing. I'd remove it for now, but can you clarify the intended behavior of `expect_column_values_to_be_equal` for the following three rows under each of the `keep_missing` settings? Would 'compare_missing' or 'include_missing' be better?  
A: [NaN, 1, 1]
B: [NaN, NaN, 1]
 
3. output_format='COMPLETE' isn't working...I know we're refactoring that and I haven't dug through yet, so just flagging in case you've thought of that specifically.

**Comment by abegong** [comment]
On 1: Good call. I agree that `expect_column_pair_values_A_to_be_greater_than_B` is better.

On 3: Hm. I didn't specifically test for this. I'll look into it.


On 2: Here's the reason we need something like `keep_missing`:

```
A = [ 1, 2, 3, 4, 5, 6,NA,NA,NA,NA ]
B = [ 1, 2, 3,-4,NA,NA,-7,-8,-9,NA ]
```

If we `expect_column_pair_values_to_be_equal`, then is this expectation 75% correct? 33.333% correct? Something else?

For single-column expectations, we've been dropping NaNs by default. This has made sense in most cases.

But if we drop any column with a NaN here, we'd get 75% correct for this expectation. This is `keep_missing="both"`. (I agree that this language is confusing.)

OTOH, we could `keep_missing="either"`, and retain all columns with at least one non-null value (all but the last entry in both A and B.) In that case, we get 3/9 = 33.333% correct.

Another alternative would be to include all NAs, which would give us 4/10 = 40% correct. (The successfully matched values are 1, 2, 3, and the NAs in the last entry.)

I don't have a good sense for which behavior will make the most sense. I suspect that the right behavior is going to differ by use case, which is why I included the `keep_missing` flag.

I also suspect this is going to be a problem for many column_pair expectations.

Thoughts?

**Comment by jpcampb2** [comment]
What if instead of keep_missing, we called it NA_matches_anything or something. That way the logic is clearer: if a value is None, it can match anything. If it's not, it matches nothing except itself.

That doesn't directly address the NaN=?NaN question: for that I think the right behavior is to drop (NaN/NaN) pairs, following the same logic as in single column expectations. If you want to reason about how many of such pairs you have, you would use the expect_column_pair_values_not_to_be_in_set('a','b', {None, None}) logic.

**Comment by jpcampb2** [comment]
@abegong ping on this...I think it was basically missing just the rename of an expectation and the NA_matches_anything or other name for the logical reasoning morass.

(also will need an update, such as including these tests into the not-implemented-in-sqlalchemy list)

**Comment by abegong** [comment]
Yes, it's the logical reasoning morass that's been slowing me down.

I'll come back to this soon.

**Comment by abegong** [comment]
This was basically a naming convention problem. I'm thinking we just use super-explicit names that avoid double negatives.

@jpcampb2, Would this work?

```
expect_column_pair_values_A_to_be_greater_than_B(
    "my_column",
    "my_other_column",
    drop_nulls_if="both_columns_are_null" / "either_column_is_null" / "never"
)
```

By default, `drop_nulls_if="both_columns_are_null"`

**Comment by jpcampb2** [comment]
I agree--here explicit is what solves the problem. After the logic trouble in the previous expect_column_values_to_[not_]be_null, I'm struggling a bit with what is the best way to be precise in language ("drop" vs "ignore" and "null" vs "missing"). 

So, my vote is for a small tweak:

```
expect_column_pair_values_A_to_be_greater_than_B(
    "my_column",
    "my_other_column",
    ignore_pair_if="both_columns_are_missing" / "either_column_is_missing" / "never"
)
```

**Comment by abegong** [comment]
I like yours better.

How about `ignore_row_if="both_values_are_missing" / "either_value_is_missing" / "never"`?

**Comment by jpcampb2** [comment]
_row_ works for me...probably nicely more general if we want to reuse the parameter across different expectations as well.

**Comment by abegong** [comment]
Okay, I can implement that.

File: great_expectations/dataset/pandas_dataset.py [hunk 1]
```diff
@@ -84,6 +84,80 @@ def inner_wrapper(self, column, mostly=None, result_format=None, *args, **kwargs
...
+        """
+        The column_pair_map_expectation decorator handles boilerplate issues surrounding the common pattern of evaluating
+        truthiness of some condition on a per row basis across a pair of columns.
+        """
+
+        @cls.expectation(inspect.getargspec(func)[0][1:])
+        @wraps(func)
+        def inner_wrapper(self, column_A, column_B, mostly=None, ignore_row_if="both_values_are_missing", result_format=None, *args, **kwargs):
+
+            if result_format is None:
```
Comment (jpcampb2) [inline great_expectations/dataset/pandas_dataset.py:98]: Use `parse_result_format` on the result here.

File: great_expectations/dataset/base.py [hunk 2]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
...
+        Args:
+            column_A (str): The first column name
+            column_B (str): The second column name
+            value_pairs_set (list of lists): All the valid pairs to be matched
+
+        Keyword Args:
+            keep_missing (str): "either" or "both"
+
+        Other Parameters:
+            output_format (str or None): \
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: Needs to get updated to new language.

File: great_expectations/dataset/base.py [hunk 3]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
...
+
+        Keyword Args:
+            allow_cross_type_comparisons (boolean or None) : If True, allow comparisons between types (e.g. integer and\
+                string). Otherwise, attempting such comparisons will raise an exception.
+
+        Keyword Args:
+            keep_missing (str): "either" or "both"
+
+        Other Parameters:
+            output_format (str or None): \
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: Needs update to the new language (just bring across all the boilerplate)

File: great_expectations/dataset/base.py [hunk 4]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
...
+
+        Args:
+            column_A (str): The first column name
+            column_B (str): The second column name
+
+        Keyword Args:
+            keep_missing (str): "either" or "both"
+
+        Other Parameters:
+            output_format (str or None): \
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: Needs update to new language (result_format)

File: great_expectations/dataset/base.py [hunk 5]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
 
         """
         raise NotImplementedError
+
+    ### Column pairs ###
+
+    def expect_column_pair_values_to_be_equal(self,
+        column_A,
+        column_B,
+        keep_missing="either",
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: Needs update to new ignore_row_if name

File: great_expectations/dataset/base.py [hunk 6]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
...
+        """
+        raise NotImplementedError
+
+    def expect_column_pair_values_A_to_be_greater_than_B(self,
+        column_A,
+        column_B,
+        or_equal=None,
+        parse_strings_as_datetimes=None,
+        allow_cross_type_comparisons=None,
+        keep_missing="either",
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: needs update to new ignore_row_if name.

Also: eventually we may want to remove parse_strings_as_datetimes option from base, since in a typed database system it's not clear we want to support in-expectation type changes. Not for this PR.

File: great_expectations/dataset/base.py [hunk 7]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
...
+                For more detail, see :ref:`meta`.
+
+        Returns:
+            A JSON-serializable expectation result object.
+
+            Exact fields vary depending on the values passed to :ref:`output_format <output_format>` and
+            :ref:`include_config`, :ref:`catch_exceptions`, and :ref:`meta`.
+
+        Notes:
+            Missing values are tricky in `column_pair` expectations. By default, `expect_column_pair_values_to_be_equal` compares
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: Needs revision to reflect new name...which I think also is clear enough that it's no longer tricky (a good thing!)

File: great_expectations/dataset/base.py [hunk 8]
```diff
@@ -3010,3 +3010,152 @@ def expect_column_kl_divergence_to_be_less_than(self,
...
```
Comment (jpcampb2) [inline great_expectations/dataset/base.py]: same as above.

File: great_expectations/dataset/pandas_dataset.py [hunk 9]
```diff
@@ -84,6 +84,80 @@ def inner_wrapper(self, column, mostly=None, result_format=None, *args, **kwargs
...
+
+            success, percent_success = self._calc_map_expectation_success(success_count, nonnull_count, mostly)
+
+            return_obj = self._format_column_map_output(
+                result_format, success,
+                element_count, nonnull_count,
+                unexpected_list, unexpected_index_list
+            )
+
+            # return_obj = self._format_column_map_output(
```
Comment (jpcampb2) [inline great_expectations/dataset/pandas_dataset.py]: Ready to pull the legacy version for the PR?

File: great_expectations/dataset/pandas_dataset.py [hunk 10]
```diff
@@ -1313,5 +1387,70 @@ def expect_column_kl_divergence_to_be_less_than(self, column, partition_object=N
...
+        if parse_strings_as_datetimes:
+            temp_column_A = column_A.map(parse)
+            temp_column_B = column_B.map(parse)
+
+        else:
+            temp_column_A = column_A
+            temp_column_B = column_B
+
+
+        if or_equal:
```
Comment (jpcampb2) [inline great_expectations/dataset/pandas_dataset.py]: or_equal == True, to be explicit since default is None? Test for nonsense values there and warn?

File: great_expectations/dataset/pandas_dataset.py [hunk 11]
```diff
@@ -1313,5 +1387,70 @@ def expect_column_kl_divergence_to_be_less_than(self, column, partition_object=N
...
+    def expect_column_pair_values_A_to_be_greater_than_B(self,
+        column_A,
+        column_B,
+        or_equal=None,
+        parse_strings_as_datetimes=None,
+        allow_cross_type_comparisons=None,
+        ignore_row_if="both_values_are_missing",
+        output_format=None, include_config=False, catch_exceptions=None
+    ):
+        #FIXME: Implement allow_cross_type_comparisons
```
Comment (jpcampb2) [inline great_expectations/dataset/pandas_dataset.py]: For now, raise ValueError? Or maybe better yet, don't have this option unless it's definitely coming soon...(it's definitely not ubiquitous in expectations)

File: great_expectations/dataset/util.py [hunk 12]
```diff
@@ -115,11 +116,21 @@ def recursively_convert_to_json_serializable(test_obj):
...
+        pass
+    except ValueError:
+        pass
+
     if isinstance(test_obj, (string_types, integer_types, float, bool)):
         # No problem to encode json
         return test_obj
 
-    elif test_obj is None:
+    if isinstance(test_obj, (str, int, float, bool)):
```
Comment (jpcampb2) [inline great_expectations/dataset/util.py]: This seems wrong...it's redundant with the check above.

File: great_expectations/dataset/util.py [hunk 13]
```diff
@@ -143,6 +154,12 @@ def recursively_convert_to_json_serializable(test_obj):
         ## to the number of digits for which the string representation will equal the float representation
```
Comment (jpcampb2) [inline great_expectations/dataset/util.py:150]: This comment (not changed here) seems wrong: shouldn't 120/131 have an explicit check on precision?

File: great_expectations/dataset/util.py [hunk 14]
```diff
@@ -143,6 +154,12 @@ def recursively_convert_to_json_serializable(test_obj):
...
         return [recursively_convert_to_json_serializable(x) for x in test_obj.tolist()]
 
+    #Note: This clause has to come after checking for np.ndarray or we get:
+    #      `ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`
```
Comment (jpcampb2) [inline great_expectations/dataset/util.py:154]: I'm confused at this change/comment...is the isinstance() call raising that ValueError?

File: tests/test_pandas_dataset.py [hunk 15]
```diff
@@ -12,6 +12,27 @@
 
 class TestPandasDataset(unittest.TestCase):
 
+    def run_encapsulated_test(self, expectation_name, filename):
```
Comment (jpcampb2) [inline tests/test_pandas_dataset.py:15]: Is this a legacy test runner sneaking back in? Can't all of these be run using the new test runner?

**Review by jpcampb2 — CHANGES_REQUESTED** [review]
Super excited about these. I left detailed inline comments already. One high-level thought: ought we try to get sql parity for these from the get-go?

File: great_expectations/dataset/pandas_dataset.py (hunk 9 above)
Comment (abegong) [reply great_expectations/dataset/pandas_dataset.py]: I'm not sure what you mean...

File: great_expectations/dataset/pandas_dataset.py (hunk 11 above)
Comment (abegong) [reply great_expectations/dataset/pandas_dataset.py]: I'd rather raise a NotImplementedError.

We're definitely going to want `allow_cross_type_comparisons` on this expectation. I'd rather make that explicit, even if I don't have time for full implementation now.

Comment (abegong) [reply great_expectations/dataset/pandas_dataset.py]: I'd rather raise a NotImplementedError.

We're definitely going to want `allow_cross_type_comparisons` on this expectation. I'd rather make that explicit, even if I don't have time for full implementation now.

File: great_expectations/dataset/pandas_dataset.py (hunk 10 above)
Comment (abegong) [reply great_expectations/dataset/pandas_dataset.py]: Good catch.

File: great_expectations/dataset/util.py (hunk 12 above)
Comment (abegong) [reply great_expectations/dataset/util.py]: Good catch.

File: tests/test_pandas_dataset.py (hunk 15 above)
Comment (abegong) [reply tests/test_pandas_dataset.py:15]: Sorry, I should've put a note on the PR before I went to bed last night.

This is a legacy test runner, which needs to be replaced before this PR is done. The tests need to be rewritten in the process, and I ran out of time for it yesterday.

File: great_expectations/dataset/pandas_dataset.py (hunk 9 above)
Comment (jpcampb2) [reply great_expectations/dataset/pandas_dataset.py]: I meant you have older code commented out...and wanted to flag that you may be ready to remove it before this PR.

File: great_expectations/dataset/util.py (hunk 14 above)
Comment (abegong) [reply great_expectations/dataset/util.py:154]: No, it's the `test_obj == None`. If test_obj is a numpy array, it attempts to compare it to None and raises the ValueError in the comment.

File: great_expectations/dataset/util.py (hunk 13 above)
Comment (abegong) [reply great_expectations/dataset/util.py:150]: I'm not sure, but this isn't a change in this version.

I propose we open a new issue to look into it.

File: great_expectations/dataset/pandas_dataset.py (hunk 1 above)
Comment (abegong) [reply great_expectations/dataset/pandas_dataset.py:98]: Can I suggest we open that as a new issue? I'm not fully up to speed on `parse_result_format`, and it looks like it's not being applied in `column_map_expectation` either.

**Comment by abegong** [comment]
This should be ready to merge. The dip in coverage is from introducing new expectations, plus a handful of hard-to-test edge cases.

File: great_expectations/dataset/pandas_dataset.py (hunk 1 above)
Comment (jpcampb2) [reply great_expectations/dataset/pandas_dataset.py:98]: Sure...although it should, and all of the other decorators already do. I just created a PR for pandas column_map_expectation. It'll be  simple fix to add to this once this PR is in.

File: great_expectations/dataset/util.py (hunk 13 above)
Comment (jpcampb2) [reply great_expectations/dataset/util.py:150]: Opened...but then looking at the source, I think there must have been some other merge issue that made my review confusing...it does have the precision check I was expecting...but it currently is only checking a couple of specific numpy types...which we should revisit.

File: great_expectations/dataset/util.py (hunk 14 above)
Comment (jpcampb2) [reply great_expectations/dataset/util.py:154]: Got it. Weird we hadn't found that earlier.

**Review by jpcampb2 — APPROVED** [review]

File: tests/test_pandas_dataset.py (hunk 15 above)
Comment (jpcampb2) [reply tests/test_pandas_dataset.py:15]: 👍
